File: android/src/main/kotlin/com/example/flutter_libdatachannel/FlutterLibdatachannelPlugin.kt
package com.example.flutter_libdatachannel

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

class FlutterLibdatachannelPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        init {
            System.loadLibrary("flutter_libdatachannel_native")
        }
    }

    // JNI methods
    private external fun nativeSetEventHandler(handler: Any?)
    private external fun nativeCreatePeerConnection(iceServersJson: String?, disableAutoNegotiation: Int): Int
    private external fun nativeClosePeerConnection(pcId: Int)
    private external fun nativeDeletePeerConnection(pcId: Int)
    private external fun nativeSetLocalDescription(pcId: Int, type: String?): Int
    private external fun nativeSetRemoteDescription(pcId: Int, sdp: String?, type: String?): Int
    private external fun nativeAddRemoteCandidate(pcId: Int, candidate: String?, mid: String?): Int
    private external fun nativeGetLocalDescription(pcId: Int): String?
    private external fun nativeGetLocalDescriptionType(pcId: Int): String?
    private external fun nativeGetRemoteDescription(pcId: Int): String?
    private external fun nativeGetRemoteDescriptionType(pcId: Int): String?
    private external fun nativeAddTrack(pcId: Int, initJson: String?): Int
    private external fun nativeDeleteTrack(trId: Int)
    private external fun nativeSendTrackMessage(trId: Int, data: ByteArray): Int
    private external fun nativeSetH264Packetizer(trId: Int, initJson: String?): Int
    private external fun nativeSetOpusPacketizer(trId: Int, initJson: String?): Int
    private external fun nativeChainRtcpReceivingSession(trId: Int): Int
    private external fun nativeChainRtcpSrReporter(trId: Int): Int
    private external fun nativeStartRecording(trId: Int, filePath: String?, codec: Int): Int
    private external fun nativeStopRecording(trId: Int): Int
    private external fun nativeStartPlayback(trId: Int, filePath: String?, speed: Double): Int
    private external fun nativePausePlayback(trId: Int): Int
    private external fun nativeResumePlayback(trId: Int): Int
    private external fun nativeStopPlayback(trId: Int): Int

    // Called from JNI when events arrive
    @Suppress("unused")
    fun onEvent(eventJson: String) {
        val map = jsonToMap(eventJson)
        eventSink?.success(map)
    }

    // Called from JNI when binary track data arrives
    @Suppress("unused")
    fun onBinaryEvent(trId: Int, data: ByteArray) {
        val map = HashMap<String, Any>()
        map["event"] = "onTrackMessage"
        map["trId"] = trId
        map["data"] = data
        eventSink?.success(map)
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_libdatachannel")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_libdatachannel/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        nativeSetEventHandler(this)
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        nativeSetEventHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "createPeerConnection" -> {
                // Dart sends iceServers as a List<String>; serialize to the JSON
                // array string the native layer expects. disableAutoNegotiation
                // arrives as an int (1/0).
                val iceServersList = call.argument<List<String>>("iceServers") ?: emptyList()
                val iceServersJson = org.json.JSONArray(iceServersList).toString()
                val disableAutoNeg = call.argument<Int>("disableAutoNegotiation") ?: 0
                val pcId = nativeCreatePeerConnection(iceServersJson, disableAutoNeg)
                if (pcId < 0) result.error("CREATE_FAILED", "Failed to create peer connection", null)
                else result.success(pcId)
            }
            "closePeerConnection" -> {
                nativeClosePeerConnection(call.argument<Int>("pcId")!!)
                result.success(null)
            }
            "deletePeerConnection" -> {
                nativeDeletePeerConnection(call.argument<Int>("pcId")!!)
                result.success(null)
            }
            "setLocalDescription" -> {
                val ret = nativeSetLocalDescription(call.argument<Int>("pcId")!!, call.argument<String>("type"))
                if (ret < 0) result.error("SET_LOCAL_DESC_FAILED", "Failed to set local description", null)
                else result.success(null)
            }
            "setRemoteDescription" -> {
                val ret = nativeSetRemoteDescription(
                    call.argument<Int>("pcId")!!,
                    call.argument<String>("sdp"),
                    call.argument<String>("type")
                )
                if (ret < 0) result.error("SET_REMOTE_DESC_FAILED", "Failed to set remote description", null)
                else result.success(null)
            }
            "addRemoteCandidate" -> {
                val ret = nativeAddRemoteCandidate(
                    call.argument<Int>("pcId")!!,
                    call.argument<String>("candidate"),
                    call.argument<String>("mid")
                )
                if (ret < 0) result.error("ADD_CANDIDATE_FAILED", "Failed to add remote candidate", null)
                else result.success(null)
            }
            "getLocalDescription" -> {
                val pcId = call.argument<Int>("pcId")!!
                val sdp = nativeGetLocalDescription(pcId)
                val type = nativeGetLocalDescriptionType(pcId)
                result.success(mapOf("sdp" to (sdp ?: ""), "type" to (type ?: "")))
            }
            "getRemoteDescription" -> {
                val pcId = call.argument<Int>("pcId")!!
                val sdp = nativeGetRemoteDescription(pcId)
                val type = nativeGetRemoteDescriptionType(pcId)
                result.success(mapOf("sdp" to (sdp ?: ""), "type" to (type ?: "")))
            }
            "addTrack" -> {
                val pcId = call.argument<Int>("pcId")!!
                val init = call.argument<Map<String, Any>>("init")
                val initJson = JSONObject(init ?: emptyMap<String, Any>()).toString()
                val trId = nativeAddTrack(pcId, initJson)
                if (trId < 0) result.error("ADD_TRACK_FAILED", "Failed to add track", null)
                else result.success(trId)
            }
            "deleteTrack" -> {
                nativeDeleteTrack(call.argument<Int>("trId")!!)
                result.success(null)
            }
            "sendTrackMessage" -> {
                val trId = call.argument<Int>("trId")!!
                val data = call.argument<ByteArray>("data")
                if (data == null) {
                    result.error("INVALID_DATA", "No data provided", null)
                } else {
                    val ret = nativeSendTrackMessage(trId, data)
                    if (ret < 0) result.error("SEND_FAILED", "Failed to send track message", null)
                    else result.success(null)
                }
            }
            "setH264Packetizer" -> {
                val trId = call.argument<Int>("trId")!!
                val init = call.argument<Map<String, Any>>("init")
                val initJson = JSONObject(init ?: emptyMap<String, Any>()).toString()
                val ret = nativeSetH264Packetizer(trId, initJson)
                if (ret < 0) result.error("SET_PACKETIZER_FAILED", "Failed to set H264 packetizer", null)
                else result.success(null)
            }
            "setOpusPacketizer" -> {
                val trId = call.argument<Int>("trId")!!
                val init = call.argument<Map<String, Any>>("init")
                val initJson = JSONObject(init ?: emptyMap<String, Any>()).toString()
                val ret = nativeSetOpusPacketizer(trId, initJson)
                if (ret < 0) result.error("SET_PACKETIZER_FAILED", "Failed to set Opus packetizer", null)
                else result.success(null)
            }
            "chainRtcpReceivingSession" -> {
                val ret = nativeChainRtcpReceivingSession(call.argument<Int>("trId")!!)
                if (ret < 0) result.error("CHAIN_FAILED", "Failed to chain RTCP receiving session", null)
                else result.success(null)
            }
            "chainRtcpSrReporter" -> {
                val ret = nativeChainRtcpSrReporter(call.argument<Int>("trId")!!)
                if (ret < 0) result.error("CHAIN_FAILED", "Failed to chain RTCP SR reporter", null)
                else result.success(null)
            }
            "startRecording" -> {
                val trId = call.argument<Int>("trId")!!
                val filePath = call.argument<String>("filePath")
                val codec = call.argument<Int>("codec") ?: 0
                val ret = nativeStartRecording(trId, filePath, codec)
                if (ret < 0) result.error("START_RECORDING_FAILED", "Failed to start recording", null)
                else result.success(null)
            }
            "stopRecording" -> {
                val trId = call.argument<Int>("trId")!!
                val ret = nativeStopRecording(trId)
                if (ret < 0) result.error("STOP_RECORDING_FAILED", "Failed to stop recording", null)
                else result.success(null)
            }
            "startPlayback" -> {
                val trId = call.argument<Int>("trId")!!
                val filePath = call.argument<String>("filePath")
                val speed = call.argument<Double>("speed") ?: 1.0
                val ret = nativeStartPlayback(trId, filePath, speed)
                if (ret < 0) result.error("START_PLAYBACK_FAILED", "Failed to start playback", null)
                else result.success(null)
            }
            "pausePlayback" -> {
                val ret = nativePausePlayback(call.argument<Int>("trId")!!)
                if (ret < 0) result.error("PAUSE_PLAYBACK_FAILED", "Failed to pause playback", null)
                else result.success(null)
            }
            "resumePlayback" -> {
                val ret = nativeResumePlayback(call.argument<Int>("trId")!!)
                if (ret < 0) result.error("RESUME_PLAYBACK_FAILED", "Failed to resume playback", null)
                else result.success(null)
            }
            "stopPlayback" -> {
                val ret = nativeStopPlayback(call.argument<Int>("trId")!!)
                if (ret < 0) result.error("STOP_PLAYBACK_FAILED", "Failed to stop playback", null)
                else result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        nativeSetEventHandler(null)
    }

    private fun jsonToMap(json: String): Map<String, Any> {
        val map = HashMap<String, Any>()
        try {
            val obj = JSONObject(json)
            for (key in obj.keys()) {
                map[key] = obj.get(key)
            }
        } catch (e: Exception) {
            // ignore parse errors
        }
        return map
    }
}

File: ios/Classes/FlutterLibdatachannelPlugin.swift
import Flutter
import UIKit

public class FlutterLibdatachannelPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private static var instance: FlutterLibdatachannelPlugin?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "flutter_libdatachannel",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "flutter_libdatachannel/events",
            binaryMessenger: registrar.messenger()
        )
        let pluginInstance = FlutterLibdatachannelPlugin()
        instance = pluginInstance
        registrar.addMethodCallDelegate(pluginInstance, channel: channel)
        eventChannel.setStreamHandler(pluginInstance)

        ldc_init()
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        setupCallbacks()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        ldc_set_event_callback(nil, nil)
        ldc_set_binary_event_callback(nil, nil)
        return nil
    }

    private func setupCallbacks() {
        let pointer = Unmanaged.passUnretained(self).toOpaque()

        ldc_set_event_callback({ (eventJson, userData) in
            guard let userData = userData,
                  let json = eventJson else { return }
            let plugin = Unmanaged<FlutterLibdatachannelPlugin>.fromOpaque(userData).takeUnretainedValue()
            let jsonStr = String(cString: json)
            if let data = jsonStr.data(using: .utf8),
               let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                DispatchQueue.main.async {
                    plugin.eventSink?(dict)
                }
            }
        }, pointer)

        ldc_set_binary_event_callback({ (trId, data, size, userData) in
            guard let userData = userData, let data = data else { return }
            let plugin = Unmanaged<FlutterLibdatachannelPlugin>.fromOpaque(userData).takeUnretainedValue()
            let flutterData = FlutterStandardTypedData(bytes: Data(bytes: data, count: Int(size)))
            let map: [String: Any] = [
                "event": "onTrackMessage",
                "trId": Int(trId),
                "data": flutterData,
            ]
            DispatchQueue.main.async {
                plugin.eventSink?(map)
            }
        }, pointer)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            if call.method == "createPeerConnection" {
                let pcId = ldc_create_peer_connection(nil, 0)
                result(Int(pcId))
                return
            }
            result(FlutterMethodNotImplemented)
            return
        }

        switch call.method {
        case "createPeerConnection":
            let iceServers = args["iceServers"] as? String
            let disableAutoNeg = args["disableAutoNegotiation"] as? Int ?? 0
            let pcId = ldc_create_peer_connection(iceServers, Int32(disableAutoNeg))
            if pcId < 0 {
                result(FlutterError(code: "CREATE_FAILED", message: "Failed to create peer connection", details: nil))
            } else {
                result(Int(pcId))
            }

        case "closePeerConnection":
            ldc_close_peer_connection(Int32(args["pcId"] as! Int))
            result(nil)

        case "deletePeerConnection":
            ldc_delete_peer_connection(Int32(args["pcId"] as! Int))
            result(nil)

        case "setLocalDescription":
            let pcId = Int32(args["pcId"] as! Int)
            let type = args["type"] as? String
            let ret = ldc_set_local_description(pcId, type)
            if ret < 0 {
                result(FlutterError(code: "SET_LOCAL_DESC_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "setRemoteDescription":
            let pcId = Int32(args["pcId"] as! Int)
            let sdp = args["sdp"] as? String
            let type = args["type"] as? String
            let ret = ldc_set_remote_description(pcId, sdp, type)
            if ret < 0 {
                result(FlutterError(code: "SET_REMOTE_DESC_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "addRemoteCandidate":
            let pcId = Int32(args["pcId"] as! Int)
            let candidate = args["candidate"] as? String
            let mid = args["mid"] as? String
            let ret = ldc_add_remote_candidate(pcId, candidate, mid)
            if ret < 0 {
                result(FlutterError(code: "ADD_CANDIDATE_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "getLocalDescription":
            let pcId = Int32(args["pcId"] as! Int)
            let sdp = ldc_get_local_description(pcId)
            let type = ldc_get_local_description_type(pcId)
            let desc: [String: String] = [
                "sdp": sdp.map { String(cString: $0) } ?? "",
                "type": type.map { String(cString: $0) } ?? "",
            ]
            if let sdp = sdp { ldc_free(sdp) }
            if let type = type { ldc_free(type) }
            result(desc)

        case "getRemoteDescription":
            let pcId = Int32(args["pcId"] as! Int)
            let sdp = ldc_get_remote_description(pcId)
            let type = ldc_get_remote_description_type(pcId)
            let desc: [String: String] = [
                "sdp": sdp.map { String(cString: $0) } ?? "",
                "type": type.map { String(cString: $0) } ?? "",
            ]
            if let sdp = sdp { ldc_free(sdp) }
            if let type = type { ldc_free(type) }
            result(desc)

        case "addTrack":
            let pcId = Int32(args["pcId"] as! Int)
            let initMap = args["init"] as? [String: Any] ?? [:]
            let initJson = mapToJson(initMap)
            let trId = ldc_add_track(pcId, initJson)
            if trId < 0 {
                result(FlutterError(code: "ADD_TRACK_FAILED", message: "Failed", details: nil))
            } else {
                result(Int(trId))
            }

        case "deleteTrack":
            ldc_delete_track(Int32(args["trId"] as! Int))
            result(nil)

        case "sendTrackMessage":
            let trId = Int32(args["trId"] as! Int)
            if let data = args["data"] as? FlutterStandardTypedData {
                let bytes = [UInt8](data.data)
                let ret = ldc_send_track_message(trId, bytes, Int32(bytes.count))
                if ret < 0 {
                    result(FlutterError(code: "SEND_FAILED", message: "Failed", details: nil))
                } else {
                    result(nil)
                }
            } else {
                result(FlutterError(code: "INVALID_DATA", message: "No data", details: nil))
            }

        case "setH264Packetizer":
            let trId = Int32(args["trId"] as! Int)
            let initMap = args["init"] as? [String: Any] ?? [:]
            let ret = ldc_set_h264_packetizer(trId, mapToJson(initMap))
            if ret < 0 {
                result(FlutterError(code: "SET_PACKETIZER_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "setOpusPacketizer":
            let trId = Int32(args["trId"] as! Int)
            let initMap = args["init"] as? [String: Any] ?? [:]
            let ret = ldc_set_opus_packetizer(trId, mapToJson(initMap))
            if ret < 0 {
                result(FlutterError(code: "SET_PACKETIZER_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "startRecording":
            let trId = Int32(args["trId"] as! Int)
            let filePath = args["filePath"] as? String
            let codec = Int32(args["codec"] as? Int ?? 0)
            let ret = ldc_start_recording(trId, filePath, codec)
            if ret < 0 {
                result(FlutterError(code: "START_RECORDING_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "stopRecording":
            let trId = Int32(args["trId"] as! Int)
            let ret = ldc_stop_recording(trId)
            if ret < 0 {
                result(FlutterError(code: "STOP_RECORDING_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "getSelectedCandidatePair":
            let pcId = Int32(args["pcId"] as! Int)
            if let cstr = ldc_get_selected_candidate_pair(pcId) {
                let s = String(cString: cstr)
                ldc_free(cstr)
                result(s)
            } else {
                result(nil)
            }

        case "chainRtcpReceivingSession":
            let ret = ldc_chain_rtcp_receiving_session(Int32(args["trId"] as! Int))
            if ret < 0 {
                result(FlutterError(code: "CHAIN_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        case "chainRtcpSrReporter":
            let ret = ldc_chain_rtcp_sr_reporter(Int32(args["trId"] as! Int))
            if ret < 0 {
                result(FlutterError(code: "CHAIN_FAILED", message: "Failed", details: nil))
            } else {
                result(nil)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func mapToJson(_ map: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: map),
              let str = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return str
    }
}
